### PR TITLE
Update qboost.py

### PR DESCRIPTION
**Pull request by PatrickHuembeli**
We added new classes to QBoost that enable QBoost Regression and QBoostPlus Regression. If you want to have an example file, go to:
https://gitlab.com/PatrickHuembeli/qml_hackathon

### CODE REVIEW

File: qboost/qboost.py [hunk 1]
```diff
@@ -13,10 +13,23 @@
...
+def weight_penalty(h, y, percent = 0.1): 
+    """
+    Penalize differences of the prediction from the label.
+    """
+    diff = np.abs(h-y)
+    minn = diff.min()
+    maxx = diff.max()
+    norm = (diff-minn)/(maxx-minn)
+    norm = 1.0*(norm  < percent)
+#     h, y = np.expm1(h), np.expm1(y)
```
Comment (mstechly) [inline qboost/qboost.py]: I suppose this comment is unnecessary

File: qboost/qboost.py [hunk 2]
```diff
@@ -13,10 +13,23 @@
...
+from sklearn.ensemble import AdaBoostClassifier, RandomForestClassifier, AdaBoostRegressor
 import numpy as np
 from copy import deepcopy
 
+def weight_penalty(h, y, percent = 0.1): 
+    """
+    Penalize differences of the prediction from the label.
+    """
+    diff = np.abs(h-y)
+    minn = diff.min()
```
Comment (mstechly) [inline qboost/qboost.py]: Why naming it minn and maxx and not min and max?

File: qboost/qboost.py [hunk 3]
```diff
@@ -13,10 +13,23 @@
 #    limitations under the License.
 
 
-from sklearn.tree import DecisionTreeClassifier
+from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor
+from sklearn.ensemble import AdaBoostClassifier, RandomForestClassifier, AdaBoostRegressor
 import numpy as np
 from copy import deepcopy
 
+def weight_penalty(h, y, percent = 0.1):
```
Comment (mstechly) [inline qboost/qboost.py]: y is a clear variable name, though h is not.
Could you use a more descriptive name? Unless this is a convention used in the whole library.

File: qboost/qboost.py [hunk 4]
```diff
@@ -149,6 +152,146 @@ def predict(self, X):
...
+        self.var1 = 0.0
+        self.qij = 0.0
+        self.norm = 0.0
+
+    def __construct_wc(self):
+
+        self.estimators_ = [DecisionTreeRegressor(max_depth=self.max_depth,
+                                                   random_state=np.random.randint(1000000,10000000))
+                            for _ in range(self.n_estimators)]
+#        self.estimators_ = [AdaBoostRegressor(random_state=np.random.randint(1000000,10000000))
```
Comment (mstechly) [inline qboost/qboost.py]: This comment is unnecessary

File: qboost/qboost.py [hunk 5]
```diff
@@ -149,6 +152,146 @@ def predict(self, X):
...
+        fit estimators
+        :param X:
+        :param y:
+        :return:
+        """
+
+        self.estimator_weights = np.zeros(self.n_estimators)
+
+        d = np.ones(len(X)) / len(X)
+        for i, h in enumerate(self.estimators_):
```
Comment (mstechly) [inline qboost/qboost.py]: I don't know what "h" means, 'estimator' or 'current_estimator' would be more descriptive name.

File: qboost/qboost.py [hunk 6]
```diff
@@ -149,6 +152,146 @@ def predict(self, X):
...
+            print(i)
```
Comment (mstechly) [inline qboost/qboost.py]: Have you considered adding "verbose" as a parameter for this class? And putting print into "if verbose:" blocks?

File: qboost/qboost.py [hunk 7]
```diff
@@ -149,6 +152,146 @@ def predict(self, X):
...
+            h.fit(X, y, sample_weight=d)
+            pred = h.predict(X)
+            norm = weight_penalty(pred, y)
+            self.norm = norm
```
Comment (mstechly) [inline qboost/qboost.py]: You are overwriting self.norm with a new norm every iteration. Is this behavior intended?

File: qboost/qboost.py [hunk 8]
```diff
@@ -149,6 +152,146 @@ def predict(self, X):
...
+        :param X:
+        :return:
+        """
+
+        if not hasattr(self, 'estimator_weights'):
+            raise Exception('Not Fitted Error!')
+
+        y = np.zeros(len(X))
+
+        for (h, w) in zip(self.estimators_, self.estimator_weights):
```
Comment (mstechly) [inline qboost/qboost.py]: Same as before - h is not a greate name.

File: qboost/qboost.py [hunk 9]
```diff
@@ -149,6 +152,146 @@ def predict(self, X):
...
+            y += w * h.predict(X)
+
+        y = np.sign(y)
+
+        return y
+
+    def copy(self):
+
+        clf = WeakRegressor(n_estimators=self.n_estimators, max_depth=self.max_depth)
```
Comment (mstechly) [inline qboost/qboost.py]: It took me several seconds to figure out why you named it clf.
'classifier' is also a good variable name, and it's not a long one ;)

File: qboost/qboost.py [hunk 10]
```diff
@@ -149,6 +152,146 @@ def predict(self, X):
...
+        clf.estimators_ = deepcopy(self.estimators_)
+        if hasattr(self, 'estimator_weights'):
+            clf.estimator_weights = np.array(self.estimator_weights)
+
+        return clf
+
+
+class QBoostRegressor(WeakRegressor):
+    """
+    Qboost
```
Comment (mstechly) [inline qboost/qboost.py]: This class description isn't very informative.

File: qboost/qboost.py [hunk 11]
```diff
@@ -149,6 +152,146 @@ def predict(self, X):
...
+    def fit(self, X, y, sampler, lmd=0.2, **kwargs):
+
+        n_data = len(X)
+
+        # step 1: fit weak classifiers
+        super(QBoostRegressor, self).fit(X, y)
+
+        # step 2: create QUBO
+        hij = []
+        for h in self.estimators_:
```
Comment (mstechly) [inline qboost/qboost.py]: Same as before.

File: qboost/qboost.py [hunk 12]
```diff
@@ -149,6 +152,146 @@ def predict(self, X):
...
+        hij = np.array(hij)
+        # scale hij to [-1/N, 1/N]
+        hij = 1. * hij / self.n_estimators
+        self.hij = hij
+        ## Create QUBO
+        qii = n_data * 1. / (self.n_estimators ** 2) + lmd - 2 * np.dot(hij, y)
+        self.var1 = qii
+        qij = np.dot(hij, hij.T)
+        self.qij = qij
+        Q = dict()
```
Comment (mstechly) [inline qboost/qboost.py]: I have no idea what Q is. maybe a better name?

File: qboost/qboost.py [hunk 13]
```diff
@@ -201,4 +344,59 @@ def predict(self, X):
...
+            hij.append(h.predict(X))
+
+        hij = np.array(hij)
+        # scale hij to [-1/N, 1/N]
+        hij = 1. * hij / self.n_estimators
+
+        ## Create QUBO
+        qii = n_data * 1. / (self.n_estimators ** 2) + lmd - 2 * np.dot(hij, y)
+        qij = np.dot(hij, hij.T)
+        Q = dict()
```
Comment (mstechly) [inline qboost/qboost.py]: Same as before

File: qboost/qboost.py (hunk 2 above)
Comment (arcondello) [reply qboost/qboost.py]: I believe that it's because `min` and `max` are built-in functions. I cannot find the relevent PEP but I believe that the preferred name mangling is `min_` and `max_`

**Comment by arcondello** [comment]
Hi @PatrickHuembeli , thanks for the submission and it was fun working with you last week. We're beginning the review process today though it looks like @mstechly has made some good suggestions.

File: qboost/qboost.py (hunk 3 above)
Comment (aidanproy) [inline qboost/qboost.py]: This function doesn't seem to be penalizing the differences between prediction and label. Instead it seems to penalize the number of differences that are in the lowest 10% of the range. Maybe you could add some comments explaining what's happening.

File: qboost/qboost.py [hunk 14]
```diff
@@ -149,6 +152,146 @@ def predict(self, X):
...
+        self.n_estimators = n_estimators
+        self.estimators_ = []
+        self.max_depth = max_depth
+        self.__construct_wc()
+        self.Qu = 0.0
+        self.hij = 0.0
+        self.var1 = 0.0
+        self.qij = 0.0
+        self.norm = 0.0
+
```
Comment (aidanproy) [inline qboost/qboost.py]: The Qu, hij, var1, and qij variables should be in the QBoostRegressor class rather than the WeakRegressor class.

File: qboost/qboost.py [hunk 15]
```diff
@@ -149,6 +152,146 @@ def predict(self, X):
         return y
 
 
+class WeakRegressor(object):
+    """
+    Weak Classifiers based on DecisionTree
+    """
+
```
Comment (aidanproy) [inline qboost/qboost.py]: This comment is confusing, should be fixed to "Weak Regressors". (This happens in other comments below as well.)

File: qboost/qboost.py (hunk 3 above)
Comment (PatrickHuembeli) [reply qboost/qboost.py]: I will add some comments there. And I honestly have to admit that this is the part of the regression code that can be improved for sure.
The idea behind this metric is simply the following:
For classification we can implement a binary penalty for 'failed' classification. We do not have this option for Regression. So our idea was to call a prediction failed if we are more than 10% from the target value. This number is arbitrary and might strongly depend on the problem. One could also think about different metrics and start exploring their performance.

File: qboost/qboost.py (hunk 4 above)
Comment (PatrickHuembeli) [reply qboost/qboost.py]: I kept it there to have an example for another possible Regressor.

File: qboost/qboost.py (hunk 14 above)
Comment (PatrickHuembeli) [reply qboost/qboost.py]: I put them into the Regressor class. They can also be deleted. I mostly used them for debugging.

File: qboost/qboost.py (hunk 15 above)
Comment (PatrickHuembeli) [reply qboost/qboost.py]: I changed them.

File: qboost/qboost.py (hunk 5 above)
Comment (PatrickHuembeli) [reply qboost/qboost.py]: h was used for hypothesis. I agree it might be a bit confusing but here we just followed the syntax of the classifier. If we change it then we also have to change it in the old part of the file.

File: qboost/qboost.py (hunk 6 above)
Comment (PatrickHuembeli) [reply qboost/qboost.py]: All the prints can be deleted. Which I did. They are not necessary. It was just sometimes convenient for debugging and speed tests.

File: qboost/qboost.py (hunk 7 above)
Comment (PatrickHuembeli) [reply qboost/qboost.py]: I deleted it. Again unnecessary and just used for debugging

File: qboost/qboost.py (hunk 8 above)
Comment (PatrickHuembeli) [reply qboost/qboost.py]: I leave this decision to Dwave officials. But I agree with you, also because hypothesis is in my opinion not used very often as well.

File: qboost/qboost.py (hunk 9 above)
Comment (PatrickHuembeli) [reply qboost/qboost.py]: Totally agree and changed

File: qboost/qboost.py (hunk 10 above)
Comment (PatrickHuembeli) [reply qboost/qboost.py]: Again, agreed. I added a bit more, but I mostly followed the rest of the code as an example.

File: qboost/qboost.py (hunk 3 above)
Comment (aidanproy) [reply qboost/qboost.py]: For regression, I would have thought the natural measure of difference between prediction and label was simply the the absolute value of the difference (or that squared). 

The sklearn implementation of Adaboost for regression follows this paper: https://dl.acm.org/citation.cfm?id=657132. So that might be the thing to do here as well.

File: qboost/qboost.py (hunk 4 above)
Comment (mstechly) [reply qboost/qboost.py]: Ok, so maybe add a comment explaining why it's here?

**Comment by mstechly** [comment]
@PatrickHuembeli thanks for the changes, it looks better :)
